llvm/test/Transforms/InstCombine/pr91127.ll

```diff
@@ -0,0 +1,82 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -S -passes=instcombine < %s | FileCheck %s
+
+define i1 @test_eq1(i32 %x, i16 %y) {
+; CHECK-LABEL: define i1 @test_eq1(
+; CHECK-SAME: i32 [[X:%.*]], i16 [[Y:%.*]]) {
+; CHECK-NEXT:    [[TMP1:%.*]] = sext i16 [[Y]] to i32
+; CHECK-NEXT:    [[COND:%.*]] = icmp eq i32 [[TMP1]], [[X]]
+; CHECK-NEXT:    ret i1 [[COND]]
+;
+  %conv1 = trunc nsw i32 %x to i8
+  %conv2 = trunc nsw i16 %y to i8
+  %cond = icmp eq i8 %conv1, %conv2
+  ret i1 %cond
+}
+
+; FIXME: It is weird that we generate truncs for test_eq2, but not for test_eq1.
```

Comment:
I don't know which should be the canonical form. Any thoughts?

Reply:
I think `sext` is typically easier to work with than `trunc`.

Reply:
Either way though for the `trunc` case we should be propagating the nowrap flags: https://alive2.llvm.org/ce/z/gb4dQQ

Reply:
At least historically our position has been that ext is preferable to trunc. (It's possible that this is not the optimal choice anymore now that we have trunc nuw/nsw...)

Reply:
I think `nsw/nuw` makes `trunc` hypothetically equally easy to work with as `sext`/`zext`, but since we already have good support for ext seems straightforward to stick with what we got.